samza-core/src/main/java/org/apache/samza/storage/SideInputStorageManager.java

```diff
@@ -0,0 +1,372 @@
+   *   3. Fetches the partition metadata for each system stream and fetch the corresponding partition metadata
+   *      and populates the oldest offset for SSPs belonging to the system stream.
+   *
+   * @return a {@link Map} of {@link SystemStreamPartition} to their oldest offset.
+   */
+  private Map<SystemStreamPartition, String> getOldestOffsets() {
+    Map<SystemStreamPartition, String> oldestOffsets = new HashMap<>();
+
+    // Step 1
+    Map<SystemStream, List<SystemStreamPartition>> systemStreamToSsp = sspsToStore.keySet().stream()
+        .collect(Collectors.groupingBy(SystemStreamPartition::getSystemStream));
+
+    // Step 2
+    Map<SystemStream, SystemStreamMetadata> metadata = JavaConverters.mapAsJavaMapConverter(
+        streamMetadataCache.getStreamMetadata(
+            JavaConverters.asScalaSetConverter(systemStreamToSsp.keySet()).asScala().toSet(), false)).asJava();
+
+    // Step 3
+    metadata.forEach((systemStream, systemStreamMetadata) -> {
+        // get the partition metadata for each system stream
+        Map<Partition, SystemStreamMetadata.SystemStreamPartitionMetadata> partitionMetadata =
+          systemStreamMetadata.getSystemStreamPartitionMetadata();
+
+        // For SSPs belonging to the system stream, use the partition metadata to get the oldest offset
+        Map<SystemStreamPartition, String> offsets = systemStreamToSsp.get(systemStream).stream()
+          .collect(
+              Collectors.toMap(Function.identity(), ssp -> partitionMetadata.get(ssp.getPartition()).getOldestOffset()));
+
+        oldestOffsets.putAll(offsets);
+      });
+
+    return oldestOffsets;
+  }
+
+  /**
+   * Initializes the store directories for all the stores:
+   *  1. Cleans up the directories for invalid stores.
+   *  2. Ensures that the directories exist.
+   */
+  private void initializeStoreDirectories() {
+    LOG.info("Initializing side input store directories.");
+
+    stores.keySet().forEach(storeName -> {
+        File storeLocation = getStoreLocation(storeName);
+        String storePath = storeLocation.toPath().toString();
+        if (!isValidSideInputStore(storeName)) {
+          LOG.info("Cleaning up the store directory at {} for {}", storePath, storeName);
+          FileUtil.rm(storeLocation);
+        }
+
+        if (!storeLocation.exists()) {
+          LOG.info("Creating {} as the store directory for the side input store {}", storePath, storeName);
+          storeLocation.mkdirs();
+        }
+      });
+  }
+
+  private boolean isValidSideInputStore(String storeName) {
+    File storeLocation = getStoreLocation(storeName);
+
+    return isPersistedStore(storeName)
+        && StorageManagerUtil.isStaleStore(storeLocation, OFFSET_FILE, STORE_DELETE_RETENTION_MS, clock.currentTimeMillis())
+        && StorageManagerUtil.isOffsetFileValid(storeLocation, OFFSET_FILE);
+  }
+
+  private boolean isPersistedStore(String storeName) {
```

Comment:
Just curious. Do we have any use-cases where side-input stores are not persisted on disk?